File: cpp/myLeetCode/Stack_Queue/239. Sliding Window Maximum.cpp
// 239. Sliding Window Maximum
// https://leetcode.com/problems/sliding-window-maximum/

#include <iostream>
#include <vector>
#include <deque>

std::vector<int> maxSlidingWindow(std::vector<int>& nums, int k) {
    std::vector<int> result;
    std::deque<int> dq;

    for (int i = 0; i < nums.size(); i++) {
        // 如果队列中的元素已经超出窗口的范围，需要移除队首元素
        if (!dq.empty() && dq.front() == i - k) {
            dq.pop_front();
        }

        // 移除队列尾部小于当前下标的元素，以保持队列单调递减
        while (!dq.empty() && nums[dq.back()] < nums[i]) {
            dq.pop_back();
        }

        // 将当前元素的索引加入队列尾部
        dq.push_back(i);

        // 如果已经形成一个完整的窗口，将队首元素（当前窗口的最大值）加入结果数组
        if (i >= k - 1) {
            result.push_back(nums[dq.front()]);
        }
    }

    return result;
}

int main() {
    std::vector<int> nums = {1, 3, -1, -3, 5, 3, 6, 7};
    int k = 3;

    std::vector<int> result = maxSlidingWindow(nums, k);

    std::cout << "Max sliding window: ";
    for (int num : result) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: CsharpProjects/myLeetCode/Stack_Queue/239. Sliding Window Maximum.cs
// 239. Sliding Window Maximum
// https://leetcode.com/problems/sliding-window-maximum/
using System.Collections.Generic;

public class Solution {
    public int[] MaxSlidingWindow(int[] nums, int k) {
        if (nums == null || nums.Length == 0) return new int[0];

        int[] result = new int[nums.Length - k + 1];
        LinkedList<int> deque = new LinkedList<int>();

        for (int i = 0; i < nums.Length; i++) {
            // 从deque的尾部移除小于滑动窗口左边界的索引
            if (deque.Count > 0 && deque.First.Value < i - k + 1) {
                deque.RemoveFirst();
            }

            // 从deque的尾部移除小于当前数字的所有索引
            while (deque.Count > 0 && nums[deque.Last.Value] < nums[i]) {
                deque.RemoveLast();
            }

            // 将当前元素添加到deque的尾部
            deque.AddLast(i);

            if (i >= k - 1) {
                result[i - k + 1] = nums[deque.First.Value];
            }
        }

        return result;
    }
}
public class Program {
    public static void Main() {
        Solution solution = new Solution();

        int[] nums = new int[] {1,3,-1,-3,5,3,6,7};
        int k = 3;
        int[] result = solution.MaxSlidingWindow(nums, k);

        Console.WriteLine($"Input: {string.Join(", ", nums)}");
        Console.WriteLine($"Output: {string.Join(", ", result)}");
    }
}
